presto-main/src/main/java/io/prestosql/operator/HttpPageBufferClient.java

```diff
@@ -612,6 +639,33 @@ public PagesResponse handle(Request request, Response response)
             }
         }
 
+        private void verifyChecksum(URI uri, long readChecksum, List<SerializedPage> pages)
+        {
+            // TODO identify our address, so that query failure reports both ends of communication
+            switch (dataIntegrityVerification) {
+                case NONE:
+                    if (readChecksum != NO_CHECKSUM) {
+                        throw new IllegalStateException(format("Expected checksum to be NO_CHECKSUM (0x%08x) but is 0x%08x, when reading from %s", NO_CHECKSUM, readChecksum, uri));
+                    }
+                    return;
+
+                case ABORT:
+                case RETRY:
+                    long calculatedChecksum = calculateChecksum(pages);
+                    if (readChecksum != calculatedChecksum) {
+                        String message = format("Data corruption, read checksum: 0x%08x, calculated checksum: 0x%08x, when reading from %s", readChecksum, calculatedChecksum, uri);
+                        if (dataIntegrityVerification == DataIntegrityVerification.RETRY) {
+                            throw new IllegalStateException(message);
+                        }
+                        throw new PrestoException(GENERIC_INTERNAL_ERROR, message);
+                    }
+                    return;
+
+                default:
+                    throw new IllegalStateException("Unknown DataIntegrityVerification type: " + dataIntegrityVerification);
```

Comment:
Given code above is right this will result in an retry. It should probably be an `GENERIC_INTERNAL_ERROR` here.

Reply:
checksum verification & retry is reworked, please see new code